File: src/MajorityElement.java
// https://leetcode.com/problems/majority-element
// T: O(N)
// S: O(1)

public class MajorityElement {
    public int majorityElement(int[] nums) {
        int candidate = nums[0], frequency = 0;

        for (int element : nums) {
            frequency += element == candidate ? 1 : -1;

            if (frequency == 0) {
                candidate = element;
                frequency = 1;
            }
        }

        return candidate;
    }
}

File: python/majority_element.py
from typing import List


class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        element, count = nums[0], 0
        for number in nums:
            count += 1 if number == element else -1
            if count == 0:
                element = number
                count = 1
        return element
